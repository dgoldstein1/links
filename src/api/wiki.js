import axios from "axios";

let _errOut = (s, e) => ({
  success: false,
  error: `Could not get description from wikipedia for article ${s}: ${e}`
});
/**
 * @param s {string}
 * @return JSON {
    success: bool,
    data : {
      description : string,
      img : "string URL"
    }
    error : string
  }
 **/
export function getDescription(s) {
  return _queryByTitle(s).then(r => {
    console.log(r);
    if (r.success) return Promise.resolve(r);
    // if can't find find exact title, try opensearch
    return _opensearch(s);
  });
}

// gets icon and description by trying to guess title
export function _queryByTitle(s) {
  // get extracts
  let url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
  url += `?action=query`;
  url += `&format=json`;
  url += `&prop=extracts`;
  url += `&exlimit=max`;
  url += `&explaintext`;
  url += `&exintro`;
  url += `&titles=${s}`;

  return axios
    .get(encodeURI(url))
    .then(r => {
      let pageId = Object.keys(r.data.query.pages)[0];
      pageId = parseInt(pageId);
      /*jslint eqeq: true*/
      if (!pageId || pageId === -1) return _errOut(s, "no page found");
      let extract = r.data.query.pages[pageId].extract;
      // now get images
      url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
      url += `?action=query`;
      url += `&prop=pageimages`;
      url += `&format=json`;
      url += `&pithumbsize=100`;
      url += `&pageids=${pageId}`;
      return axios
        .get(encodeURI(url))
        .then(r => ({
          success: true,
          description: extract,
          img: r.data.query.pages[pageId].thumbnail
        }))
        .catch(_errOut);
    })
    .catch(_errOut);
}

// gets description and image for the first
// result in an opensearch
export function _opensearch(s) {
  // get extracts
  let url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
  url += "?action=opensearch";
  url += "&limit=1";
  url += "&namespace=0";
  url += "&format=json";
  url += `&search=${s}`;

  return axios
    .get(encodeURI(url))
    .then(r => {
      let pageTitle = r.data[1];
      if (!pageTitle) return _errOut(s, "no page found");
      let extract = r.data[2];
      // now get images
      url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
      url += `?action=query`;
      url += `&prop=pageimages`;
      url += `&format=json`;
      url += `&pithumbsize=100`;
      url += `&titles=${pageTitle}`;
      return axios
        .get(encodeURI(url))
        .then(r => {
          let pageId = Object.keys(r.data.query.pages)[0];
          return {
            success: true,
            description: extract,
            img: pageId && r.data.query.pages[pageId].thumbnail
          };
        })
        .catch(_errOut);
    })
    .catch(_errOut);
}
