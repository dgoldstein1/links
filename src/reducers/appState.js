import { SET_LOADING, UPDATE_VIEW, SET_FATAL_ERROR } from "../actions/appState";

const initialState = {
  loading: true, // the app is / isn't loading
  fatalError: undefined, // error which  causes app to not load
  view: "splash" // one of "explore", "path", "splash", "about", "settings"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_FATAL_ERROR:
      return Object.assign({}, state, {
        fatalError: action.fatalError
      });
    // set when the app is loading
    case SET_LOADING:
      return Object.assign({}, state, {
        loading: action.loading === undefined ? false : action.loading
      });
    // update where the user is currently looking
    case UPDATE_VIEW:
      return Object.assign({}, state, {
        view: action.view
      });
    default:
      return state;
  }
};

export default visitReducer;
