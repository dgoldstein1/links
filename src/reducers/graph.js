import _ from "lodash";

const initialState = {
  rootNode: {}, // root node of graph (what is in search or 'start from')
  selectedNode: {
    // node currently in view
    loading: true,
    error: "",
    description: "",
    img: "",
    node: {}
  },
  targetNode: {}, // target node in path ('ending at..')
  graph: {
    nodes: [],
    edges: []
  },
  maxShortestPaths: 5,
  directedShortestPath: true,
  loading: false,
  error: undefined,
  layout: "hierarchy",
  maxNeighbors: 15,
  topInfo: {
    betweenessEdges: [],
    betweenessNodes: [],
    pageRank: []
  }
};

const graphReducer = (state = initialState, action) => {
  switch (action.type) {
    case "CLEAR_GRAPH":
      return Object.assign({}, state, {
        graph: {
          nodes: [],
          edges: []
        }
      });
    case "SET_DIRECTED_SHORTEST_PATH":
      return Object.assign({}, state, {
        directedShortestPath: action.directedShortestPath
      });
    case "SET_MAX_NEIGHBORS":
      return Object.assign({}, state, {
        maxNeighbors: action.maxNeighbors
      });
    case "SET_GRAPH_ERROR":
      return Object.assign({}, state, {
        error: action.error
      });
    case "SET_GRAPH_LAYOUT":
      return Object.assign({}, state, {
        layout: action.layout
      });
    case "SET_GRAPH_LOADING":
      return Object.assign({}, state, {
        loading: action.loading
      });
    case "SET_ROOT_NODE":
      return Object.assign({}, state, {
        rootNode: _generateRoot(action.node.label, action.node.id)
      });
    case "SET_MAX_SHORTEST_PATHS":
      return Object.assign({}, state, {
        maxShortestPaths: action.maxShortestPaths
      });
    case "SET_ALL_PATHS_UNIQUE":
      return Object.assign({}, state, {
        pathsAreUnique: action.pathsAreUnique
      });

    case "SET_SELECTED_NODE":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          node: action.node
        }
      });
    case "SET_SELECTED_NODE_INFO":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          description: action.description,
          img: action.img,
          loading: action.loading,
          error: action.error,
          centrality: action.centrality
        }
      });
    case "SET_TARGET_NODE":
      return Object.assign({}, state, {
        targetNode: action.node
      });

    case "SET_GRAPH_PATH":
      // create array of all nodes from entries
      let nodes = action.entries.map(e => ({
        id: e.value,
        label: e.key,
        size: 1,
        color: "red"
      }));
      // create edges from each path in graph
      let edges = [];
      let entriesReverseMap = {};
      action.entries.forEach(e => (entriesReverseMap[e.value] = e.key));
      action.paths.forEach(p => {
        p.forEach((id, i) => {
          // dont add edge if we've reached the end
          if (i === p.length - 1) return;
          // lookup this entry and next entry
          edges.push({
            id: `${id}->-->--${p[i + 1]}`,
            source: id,
            target: p[i + 1],
            color: "red"
          });
        });
      });
      return Object.assign({}, state, {
        graph: { nodes, edges }
      });
    case "ADD_NEIGHBORS_TO_GRAPH":
      // set initial position to source
      action.neighbors = action.neighbors.map(n => ({
        ...n,
        size: 1
      }));
      // get list of edges to add
      let edgesToAdd = [];
      action.neighbors.forEach(n => {
        if (!n.id) console.log(n);
        edgesToAdd.push({
          id: `${action.node.id}->-->--${n.id}`,
          source: action.node.id,
          target: n.id,
          color: "black"
        });
      });
      // add in node to make sure it's in list of nodes
      action.neighbors.push(action.node);
      // return state
      return Object.assign({}, state, {
        graph: {
          nodes: _.unionBy(state.graph.nodes, action.neighbors, "id"),
          edges: _.unionBy(state.graph.edges, edgesToAdd, "id")
        }
      });
    default:
      return state;
  }
};

export function _generateRoot(label, id) {
  return {
    label,
    id,
    x: 0,
    y: 0,
    size: 1
  };
}

export default graphReducer;
