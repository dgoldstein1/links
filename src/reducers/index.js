// redux utils
import { combineReducers, createStore, applyMiddleware } from "redux";
import { createLogger } from "redux-logger";

// reducers in this directory
import appState from "./appState";
import graph from "./graph";

const Reducer = combineReducers({
  appState,
  graph,
});

const logger = createLogger({
  collapsed: true,
  diff: true,
});

let middleware = {};
if (process.env.REACT_APP_ENV === "development") {
  middleware = applyMiddleware(logger);
}

const store = createStore(Reducer, middleware);

export { store };
