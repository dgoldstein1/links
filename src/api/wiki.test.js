import { _queryByTitle, _opensearch, getDescription } from "./wiki";
import moxios from "moxios";
import { store } from "../reducers";
import sinon from "sinon";
import axios from "axios";

describe("wiki", () => {
  describe("get description", () => {
    let testTable = [
      {
        name: "query by title",
        s: "test",
        before: () => {}
      },
      {
        name: "_opensearch",
        s: "test1",
        before: () => {
          let setup = {
            name: "there is an extract",
            s: "test1",
            url:
              "/services/wiki/w/api.php?action=query&format=json&prop=extracts&exlimit=max&explaintext&exintro&redirects&titles=test1",
            responseCode: 200,
            data: {
              batchcomplete: "",
              query: {
                pages: {
                  "668126": {
                    pageid: 668126,
                    ns: 0,
                    title: "dish",
                    extract: "a dish is a dish"
                  }
                }
              },
              limits: { extracts: 20 }
            },
            getImageUrl:
              "/services/wiki/w/api.php?action=query&prop=pageimages&format=json&pithumbsize=100&pageids=668126",
            getImageResponse: {
              batchcomplete: "",
              query: {
                pages: { "668126": { pageid: 668126, ns: 0, title: "Dish" } }
              }
            }
          };
          moxios.stubRequest(setup.url, {
            status: setup.responseCode,
            response: setup.data
          });
          // mock get image, if any
          moxios.stubRequest(setup.getImageUrl, {
            status: 200,
            response: setup.getImageResponse
          });
        }
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        t.before();
        getDescription(t.s).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
  describe("_opensearch", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "success",
        s: "dish",
        searchUrl:
          "/services/wiki/w/api.php?action=opensearch&limit=1&namespace=0&format=json&search=dish",
        searchResponse: [
          "al-helal",
          ["Al-Helal Islami Academy & College"],
          [
            "Al-Helal Islami Academy & College (Bengali: \u0986\u09b2-\u09b9\u09c7\u09b2\u09be\u09b2 \u0987\u09b8\u09b2\u09be\u09ae\u09c0 \u098f\u0995\u09be\u09a1\u09c7\u09ae\u09bf \u098f\u09cd\u09af\u09be\u09a8\u09cd\u09a1 \u0995\u09b2\u09c7\u099c) is a private higher secondary school in Sapahar, Naogaon District, Bangladesh."
          ],
          ["https://en.wikipedia.org/wiki/Al-Helal_Islami_Academy_%26_College"]
        ],
        searchStatusCode: 200,
        getImageUrl:
          "/services/wiki/w/api.php?action=query&prop=pageimages&format=json&pithumbsize=100&titles=Al-Helal%20Islami%20Academy%20&%20College",
        getImageResponse: {
          batchcomplete: "",
          query: {
            pages: { "668126": { pageid: 668126, ns: 0, title: "Dish" } }
          }
        }
      },
      {
        name: "no page found",
        s: "dish",
        searchUrl:
          "/services/wiki/w/api.php?action=opensearch&limit=1&namespace=0&format=json&search=dish",
        searchResponse: [],
        searchStatusCode: 200
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.searchUrl, {
          status: t.responseCode,
          response: t.searchResponse
        });
        // mock get image, if any
        moxios.stubRequest(t.getImageUrl, {
          status: 200,
          response: t.getImageResponse
        });
        _opensearch(t.s).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
  describe("_queryByTitle", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "there is an extract",
        s: "test1",
        url:
          "/services/wiki/w/api.php?action=query&format=json&prop=extracts&exlimit=max&explaintext&exintro&redirects&titles=test1",
        responseCode: 200,
        data: {
          batchcomplete: "",
          query: {
            pages: {
              "668126": {
                pageid: 668126,
                ns: 0,
                title: "dish",
                extract: "a dish is a dish"
              }
            }
          },
          limits: { extracts: 20 }
        },
        getImageUrl:
          "/services/wiki/w/api.php?action=query&prop=pageimages&format=json&pithumbsize=100&pageids=668126",
        getImageResponse: {
          batchcomplete: "",
          query: {
            pages: { "668126": { pageid: 668126, ns: 0, title: "Dish" } }
          }
        }
      },
      {
        name: "500 response",
        s: "test1",
        url:
          "/services/wiki/w/api.php?action=query&format=json&prop=extracts&exlimit=max&explaintext&exintro&redirects&titles=test1",
        responseCode: 500,
        data: {
          batchcomplete: "",
          query: {
            pages: {
              "668126": {
                pageid: 668126,
                ns: 0,
                title: "dish",
                extract: "a dish is a dish"
              }
            }
          },
          limits: { extracts: 20 }
        },
        getImageUrl:
          "/services/wiki/w/api.php?action=query&prop=pageimages&format=json&pithumbsize=100&pageids=668126",
        getImageResponse: {
          batchcomplete: "",
          query: {
            pages: { "668126": { pageid: 668126, ns: 0, title: "Dish" } }
          }
        }
      },
      {
        name: "no page id",
        s: "test",
        url:
          "/services/wiki/w/api.php?action=query&format=json&prop=extracts&exlimit=max&explaintext&exintro&redirects&titles=test",
        responseCode: 200,
        data: {
          batchcomplete: "",
          query: {
            normalized: [{ from: "pesticidal", to: "Pesticidal" }],
            pages: { "-1": { ns: 0, title: "Pesticidal", missing: "" } }
          },
          limits: { extracts: 20 }
        }
      },
      {
        name: "no extract found",
        s: "test",
        url:
          "/services/wiki/w/api.php?action=query&format=json&prop=extracts&exlimit=max&explaintext&exintro&redirects&titles=test",
        responseCode: 200,
        data: {
          batchcomplete: "",
          query: {
            pages: {
              "668126": {
                pageid: 668126,
                ns: 0,
                title: "dish"
              }
            }
          },
          limits: { extracts: 20 }
        }
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        // mock get image, if any
        moxios.stubRequest(t.getImageUrl, {
          status: 200,
          response: t.getImageResponse
        });
        _queryByTitle(t.s).then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
});
