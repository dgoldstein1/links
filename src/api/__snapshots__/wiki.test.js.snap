// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wiki _opensearch no page found 1`] = `
Object {
  "error": "Could not get description from wikipedia for article dish: no page found",
  "success": false,
}
`;

exports[`wiki _opensearch success 1`] = `
Object {
  "description": "Al-Helal Islami Academy & College (Bengali: আল-হেলাল ইসলামী একাডেমি এ্যান্ড কলেজ) is a private higher secondary school in Sapahar, Naogaon District, Bangladesh.",
  "img": undefined,
  "success": true,
}
`;

exports[`wiki _queryByTitle 500 response 1`] = `
Object {
  "error": "Could not get description from wikipedia for article Error: Request failed with status code 500: undefined",
  "success": false,
}
`;

exports[`wiki _queryByTitle no extract found 1`] = `
Object {
  "error": "Could not get description from wikipedia for article test: no extract found",
  "success": false,
}
`;

exports[`wiki _queryByTitle no page id 1`] = `
Object {
  "error": "Could not get description from wikipedia for article test: no page found",
  "success": false,
}
`;

exports[`wiki _queryByTitle there is an extract 1`] = `
Object {
  "description": "a dish is a dish",
  "img": undefined,
  "success": true,
}
`;

exports[`wiki get description _opensearch 1`] = `
Object {
  "error": "Could not get description from wikipedia for article Error: Request failed with status code 404: undefined",
  "success": false,
}
`;

exports[`wiki get description query by title 1`] = `
Object {
  "error": "Could not get description from wikipedia for article Error: Request failed with status code 404: undefined",
  "success": false,
}
`;
